crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs

```diff
@@ -251,6 +253,43 @@ impl IndexeddbCryptoStore {
         self.static_account.read().unwrap().clone()
     }
 
+    /// Transform an [`InboundGroupSession`] into a `JsValue` holding a
+    /// [`InboundGroupSessionIndexedDbObject`], ready for storing.
+    async fn serialize_inbound_group_session(
+        &self,
+        session: &InboundGroupSession,
+    ) -> Result<JsValue> {
+        let obj = InboundGroupSessionIndexedDbObject {
+            data: self.serializer.serialize_value_as_bytes(&session.pickle().await)?,
+            needs_backup: !session.backed_up(),
+        };
+        Ok(serde_wasm_bindgen::to_value(&obj)?)
+    }
+
+    /// Transform a JsValue holding a [`InboundGroupSessionIndexedDbObject`]
+    /// back into a [`InboundGroupSession`].
+    fn deserialize_inbound_group_session(
+        &self,
+        stored_value: JsValue,
+    ) -> Result<InboundGroupSession> {
+        let idb_object: InboundGroupSessionIndexedDbObject =
+            serde_wasm_bindgen::from_value(stored_value)?;
+        let pickled_session = self.serializer.deserialize_value_from_bytes(&idb_object.data)?;
+        let session = InboundGroupSession::from_pickle(pickled_session)
+            .map_err(|e| IndexeddbCryptoStoreError::CryptoStoreError(e.into()))?;
+
+        // Although a "backed up" flag is stored inside the `data`, it is not maintained
```

Comment:
It's not clear what `data` is referring to, here? (I suppose the `idb_object.data` — maybe a sign that the `data` field should be named something else)

Also: why not maintain the `backed_up` flag when resetting the backups? At a first glance, it seems that resetting the backups is the least frequent, when deserializing an inbound group session would be quite frequent. And I'm wary of adding logic to a deserialization function; ideally we would not need to have to remember that we can't just plain deserialize-and-build-from-pickle as we consistently do elsewhere.

Reply:
> It's not clear what data is referring to, here? (I suppose the idb_object.data — maybe a sign that the data field should be named something else)

Renamed, as below.

> Also: why not maintain the `backed_up` flag when resetting the backups? At a first glance, it seems that resetting the backups is the least frequent, when deserializing an inbound group session would be quite frequent. 

Well, because it would involve decrypting, deserializing, updating, re-serializing, and re-encrypting each of possibly hundreds of thousands of pickled sessions. I don't think that's a price worth paying to simplify the code a bit. (The decryption and encryption bits are particularly expensive.)

> Ideally we would not need to have to remember that we can't just plain deserialize-and-build-from-pickle as we consistently do elsewhere.

I don't disagree, but any deserialization is necessarily going to happen via `deserialize_inbound_group_session`. In other words, the awkwardness is constrained to one particular function, and I don't really see that we'd need to remember it elsewhere.